### ISSUE TITLE
Strange loss behaviour

### ISSUE THREAD

**Issue by nokados** [RaRe-Technologies/gensim#2709]
#### Problem description

I monitor the loss during training word2vec. I have a callback that shows me how the loss was changed after each epoch. Look at the picture below:
![loss](https://user-images.githubusercontent.com/5142577/71446682-16f41180-2737-11ea-870c-68e1b0873a5f.png)

The difference between "loss" and "train_loss" is that "loss" is calculated at each epoch in the callback and "train_loss" is calculated as a loss on a copy of the model in "train" method with alpha near to zero (as well as "valid_loss") on the same dataset, it is like a loss, after saving and loading model. 

I expect that "loss" and "train_loss" will be equal, but they are very different: while the "loss" decreases rapidly, the "train_loss" remains at the same level. Is it expected behaviour?

This behaviour remains the same for any values provided to `sg` and `hs` arguments.

#### Steps/code/corpus to reproduce

See this notebook. I tried to keep it as small as possible.
https://colab.research.google.com/drive/1FF9J-DVplEUCrC47Cl1WU4jJHXPStBSc

#### Versions

Linux-4.14.137+-x86_64-with-Ubuntu-18.04-bionic
Python 3.6.9 (default, Nov  7 2019, 10:44:02) 
[GCC 8.3.0]
NumPy 1.17.4
SciPy 1.3.3
gensim 3.6.0
FAST_VERSION 1

**Comment by gojomo** [comment]
Code questions that aren't necessarily bugs or feature requests are better served on the project discussion list, <https://groups.google.com/forum/#!forum/gensim>, than this issue-tracker.

Notably: 

* the running-loss-tally reported isn't quite what most people expect; see #2617, #2135 for details. 

* your code is a bit convoluted, but it appears to be running an **extra** `train()` epoch (updating the model, with an ) every time you evaluate the model (at `on_train_begin` and every `on_epoch_end`), for two different evaluation sets. So compared to the 50 epochs of training you've specified, which would be gently gliding the effective `alpha` from the default `0.025` to `0.001`, there's another interspersed **102** extra epochs you're triggering during callbacks, and each of those starts with an extra-large `alpha=0.1` before descending (inside the single epoch) to `0`. (And, I suspect this interspersed alpha-modification may foul the main loop's alpha-management.) So, it's hard to say anything about what would be "expected" in such a setup. 

I suggest: only run `train()` once, in one place. In callbacks, just read loss values out of the model – adjusting them to be "since last epoch" if necessary for your usage. And more generally, note that "loss" isn't the most important indicator of model quality - only whether the training procedure is still making progress against its optimization target. (When loss stops improving, the model isn't getting any better with more training epochs... but a model with a higher loss value may still be more useful on downstream tasks, and a model with a zero or very-low loss value may be overfit & thus fairly useless on downstream tasks.)

**Comment by gojomo** [comment]
Also note that the `WARNING`...

    2019-12-25 09:59:39,478 [WARNING] Each 'sentences' item should be a list of words (usually unicode strings). First item here is instead plain <class 'str'>.

...and tiny number of learned words (104) indicate you're passing strings into `Word2Vec`, rather than the lists-of-word-tokens it needs. So this model is only learning vectors per character (item in a string) instead of words (item in a list).

**Comment by mpenkov** [comment]
Closing as off-topic for this issue tracker.

**Comment by gauravgund** [comment]
Hi @gojomo @mpenkov @nokados @prb @moustaki ,

I am using word2vec in gensim to generate product embeddings. I am taking product IDs as words to generate embeddings. While training the model, Loss at every epoch is 0 and I am getting a warning like this:

`WARNING:gensim.models.word2vec:EPOCH - 1 : supplied example count (0) did not equal expected count (3)`

Gensim version: 4.1.2
Google colab

Due to data sensitivity, I am using a sample to data to show my code:

`
#Data 
train_list = [['cat','dog'],
 ['apple',
  'bannaa',
  'pineapple',
  'grapes',
  'mango'], ['dog', 'apple']]

# creating generator object for data

class data_prep:
  def __iter__(self):
    for i in train_list:
      yield(i)

t1 =data_prep()

model = Word2Vec( min_count=1,vector_size= 64,workers=-1, window =2, sg = 0)
model.build_vocab(t1)

class callback(CallbackAny2Vec):
    '''Callback to print loss after each epoch.'''

    def __init__(self):
        self.epoch = 0

    def on_epoch_end(self, model):
        loss = model.get_latest_training_loss()
        print('Loss after epoch {}: {}'.format(self.epoch, loss))
        self.epoch += 1

model.train(t1, total_examples=model.corpus_count,epochs=10, compute_loss = True, callbacks=[callback()])
`
The output is as follows:
WARNING:gensim.models.word2vec:EPOCH - 1 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 2 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 3 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 4 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 5 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 6 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 7 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 8 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 9 : supplied example count (0) did not equal expected count (3)
WARNING:gensim.models.word2vec:EPOCH - 10 : supplied example count (0) did not equal expected count (3)
Loss after epoch 0: 0.0
Loss after epoch 1: 0.0
Loss after epoch 2: 0.0
Loss after epoch 3: 0.0
Loss after epoch 4: 0.0
Loss after epoch 5: 0.0
Loss after epoch 6: 0.0
Loss after epoch 7: 0.0
Loss after epoch 8: 0.0
Loss after epoch 9: 0.0
(0, 0)


Could you please help me resolve this?

**Comment by gojomo** [comment]
Help might be possible, but not here: the issue tracker is for project bug reports & feature requests, not usage advice. And that's doubly-true for old, closed issues on unrelated topics! 

See the link to the project discussion forum in the above comments, or try StackOverflow.
